Memory efficient implemenation of Levenshtein Distance for radiff2

NikolaiHampton opened the pull request:
Implemented the iterative two matrix row version from here: https://en.wikipedia.org/wiki/Levenshtein_distance
and here: http://www.codeproject.com/Articles/13525/Fast-memory-efficient-Levenshtein-algorithm

Memory usage was originally 200Gig for a couple of 200K files ~(sizeA*sizeB), now about 400K ~(sizeB*2). Speed seems similar.

Could use a clean-up but that's all I have time for unfortunately..

radare commented:
weird that it seems to break the radiff2 -g test

radare commented:
There's a segfault (null deref) here:

```
 + 746 at diff.c:237
   234 		if (distance != NULL)
   235 			*distance = v1[lb];
   236 		if (similarity != NULL)
-> 237 			*similarity = (double)1 - (double)(*distance)/(double)(R_MAX(la, lb));
   238
   239
   240 		return true;
(lldb) print distance
(unsigned int *) $0 = 0x0000000000000000
(lldb)
```

Maijin commented:
5000 !!!

Maijin commented:
See the test:

```
33m[  ]  3  radiff2: [0mradiff2 -g                    
[31m[XX][0m

[35mRegression: [0mrun_tests.sh
[35mCommand: [0m /home/travis/build/radare/radare2/install/bin/radare2 -e scr.color=0 -N -q -i /tmp/r2-regressions//radiff2-tor1hZ/rad   - > /tmp/r2-regressions//radiff2-tor1hZ/out 2> /tmp/r2-regressions//radiff2-tor1hZ/err
[35mFile: [0m-
[35mScript: [0m!!radiff2 -g main ../../bins/other/radiff2/false ../../bins/other/radiff2/true~digrap
[35mDiff: [0m--- /tmp/r2-regressions//radiff2-tor1hZ/out	2016-05-25 07:15:21.117632812 +0000
+++ /tmp/r2-regressions//radiff2-tor1hZ/exp	2016-05-25 07:15:21.117632812 +0000
@@ -0,0 +1 @@
+digraph code {
```

radare commented:
there's nothing to see in the test, because @jduck broke it. the problem is a null deref.

radare commented:
Also, don't use != NULL or == NULL, this is implicit by if()

radare commented on libr/util/diff.c:
```diff
@@ -206,49 +214,28 @@ R_API bool r_diff_buffers_distance(RDiff *d, const ut8 *a, ut32 la, const ut8 *b
+	for (i = 0; i < lb+1 ; i++) v0[i]=i;
+	
+	for (i = 0; i < la; i++) {
+		v1[0]=i + 1;
```
missing spaces around the '='

radare commented on libr/util/diff.c:
```diff
@@ -206,49 +214,28 @@ R_API bool r_diff_buffers_distance(RDiff *d, const ut8 *a, ut32 la, const ut8 *b
-			m[i][j] = R_MIN (tmin, m[i-1][j-1] + cost);
+		for (j = 0; j < lb; j++) {
+			int cost = (a[i]==b[j]) ? 0 : 1;
+			int smallest=((v1[j]+1) < (v0[j+1]+1)) ? v1[j]+1 : v0[j+1]+1;
```
missing spaces around ==, = and +

radare commented on libr/util/diff.c:
```diff
@@ -206,49 +214,28 @@ R_API bool r_diff_buffers_distance(RDiff *d, const ut8 *a, ut32 la, const ut8 *b
+		for (j = 0; j < lb; j++) {
+			int cost = (a[i]==b[j]) ? 0 : 1;
+			int smallest=((v1[j]+1) < (v0[j+1]+1)) ? v1[j]+1 : v0[j+1]+1;
+			smallest=(smallest < (v0[j]+cost)) ? smallest : v0[j]+cost;
```
more missing spaces

radare commented on libr/util/diff.c:
```diff
@@ -206,49 +214,28 @@ R_API bool r_diff_buffers_distance(RDiff *d, const ut8 *a, ut32 la, const ut8 *b
+		for (j = 0; j < lb+1; j++) v0[j]=v1[j];
 	}
 
+		
```
unnecessary newline

radare commented on libr/util/diff.c:
```diff
@@ -206,49 +214,28 @@ R_API bool r_diff_buffers_distance(RDiff *d, const ut8 *a, ut32 la, const ut8 *b
+		*distance = v1[lb];
 	if (similarity != NULL)
-		*similarity = (double)1 - (double)(m[la][lb])/(double)(R_MAX(la, lb));
+		*similarity = (double)1 - (double)(*distance)/(double)(R_MAX(la, lb));
```
distance can be NULL here, not sure if nesting if (similarity) under if (distance) would be a correct fix. please confirm

radare commented on libr/util/diff.c:
```diff
@@ -206,49 +214,28 @@ R_API bool r_diff_buffers_distance(RDiff *d, const ut8 *a, ut32 la, const ut8 *b
-		*distance = m[la][lb];
+		*distance = v1[lb];
 	if (similarity != NULL)
-		*similarity = (double)1 - (double)(m[la][lb])/(double)(R_MAX(la, lb));
```
missing spaces before ( and /

radare commented on libr/util/diff.c:
```diff
@@ -206,49 +214,28 @@ R_API bool r_diff_buffers_distance(RDiff *d, const ut8 *a, ut32 la, const ut8 *b
 	if (distance != NULL)
-		*distance = m[la][lb];
+		*distance = v1[lb];
 	if (similarity != NULL)
```
unnecesary != NULL

radare commented on libr/util/diff.c:
```diff
@@ -193,9 +193,17 @@ R_API int r_diff_buffers(RDiff *d, const ut8 *a, ut32 la, const ut8 *b, ut32 lb)
+	*/
+	
+	int i, j;
+	int v0[lb+1];
```
missing spaces aorund +

jduck commented:
BTW, the binary for the radiff2 -g test does not exist. Is that expected?

jduck commented:
Ooops. my bad. I see it's "~" not "-"

NikolaiHampton commented:
I've done some cleanup. 
* I've changed how we malloc to the v0 and v1 arrays (to avoid exhausting the stack).
* I've nested the distance and similarity calculations to avoid to NULL dereference, I don't truly know the circumstances where distance and significance could be NULL, but it's in there. 
* I've added a stderr output of progress that can be activated with -V verbose for radiff2 (only works with -s distance diff), because large files can take hours to process, it's nice to see progress. Don't know what flags you guys normally use, but -V works for me!

I'm out of time (uni commitments are going nuts), so I won't be able to do any more on this. You guys can have at it if you like, otherwise revert/close/discard as you please! :)

Cheers!
Nikolai.

radare commented:
```
diff.c:117:7: error: too few arguments to function â€˜r_diff_buffers_distanceâ€™
```

don't worry, i'll take care of those details and merge. thanks for the contribution. it's really nice to have this new algorithm implemented which clearly have a performance and memory benefits

radare commented:
use eprintf instead of fprintf

NikolaiHampton commented:
Hey, sit tight. I'm just testing something.. I have a 25% speed up using a 2d array instead of copying using [ i % 2 ] and [ (i +1) % 2] to reference the previous column.. 
I'm doing large files and it's taking too long! :)
I'll add a note here tomorrow.

radare commented:
oh you mean that maybe it will be better to keep old algorithm for some cases? im cleaning up the commits and i was about to merge in few mins..

NikolaiHampton commented:
No, the old algorithm allocates N * M  so the it's a huge array..
The new one allocates M * 2 and swaps them each iteration ( N[1..x] = M[1..x] )..
I'm trying to avoid the memory copy and use ( i mod 2) to reference the left column, and ( i + 1 mod 2) to reference the right column.
Come to think of it.. Probably just easier to switch the pointers.

radare commented:
Ah :) then let me merge now so you can see the changes and you can do that change on top of that

NikolaiHampton commented on libr/util/diff.c:
```diff
@@ -206,49 +214,28 @@ R_API bool r_diff_buffers_distance(RDiff *d, const ut8 *a, ut32 la, const ut8 *b
+			v1[j+1]=smallest;
 		}
+
+		for (j = 0; j < lb+1; j++) v0[j]=v1[j];
```
This is what I'm trying to avoid.. A wasteful copy from v1 to v0, instead just change the references.

NikolaiHampton commented:
Ok.. Merge away.. I'll pull the master down after that.
Cheers!

radare commented:
cleaned up and merged! i have also updated the radiff2 manpage to include -V

radare commented:
and fixed another segfault
